src/include/enclave/lthread.h

```diff
@@ -185,11 +185,83 @@ struct lthread_sched
     struct lthread* current_lthread;
 };
 
+struct schedctx {
+	/* Part 1 -- these fields may be external or
+	 * internal (accessed via asm) ABI. Do not change. */
+	struct schedctx *self;
+	uintptr_t *dtv;
+	void *unused1, *unused2;
+	uintptr_t sysinfo;
+	uintptr_t canary, canary2;
+
+	/* Part 2 -- implementation details, non-ABI. */
```

Comment:
Note that most of these fields are there only because `schedctx` is the same layout as the TLS segment in the old implementation, so that it can live in the same place.  Most of them can be removed.  In fact, I think `lthread_sched` is the only thing that we actually need here (though the initial self pointer is also useful because x86 doesn't give you any other easy way of accessing things in the gs segment).

Reply:
Do we really need the self pointer? schedctx is located at the end of the ethread gsbase pointed page. And is accessed via a pointer placed at gs:48.

Reply:
No, if we're just storing a pointer to this in gs:48 and not the structure in gs, we don't need the self pointer.  The self pointer was there because x86 doesn't have instructions for translating from a segment-relative address to a linear address.  We were storing this structure in the FS segment and the only way of loading a pointer to it was to load the self pointer (%fs:0).  We now have a pointer to it in %gs, so we don't need the self pointer.